euphoria-spark/src/main/java/cz/seznam/euphoria/spark/ReduceByKeyTranslator.java

```diff
@@ -93,17 +104,46 @@ static boolean wantTranslate(ReduceByKey operator) {
             final OUT el = (OUT) t._2();
             return new SparkElement<>(kw.window(), kw.timestamp(), Pair.of(kw.key(), el));
           });
-    } else {
-      final JavaPairRDD<KeyedWindow<W, KEY>, OUT> reduced =
-          tuples.groupByKey().flatMapValues(new Reducer<>(reducer, accumulatorProvider));
+    }
 
-      return reduced.map(
-          t -> {
-            final KeyedWindow<W, KEY> kw = t._1();
-            final OUT el = t._2();
-            return new SparkElement<>(kw.window(), kw.timestamp(), Pair.of(kw.key(), el));
-          });
+    final JavaPairRDD<KeyedWindow<W, KEY>, OUT> reduced;
+
+    if (ClassUtils.isComparable(operator.getKeyClass())) {
+      final Partitioner partitioner = new HashPartitioner(input.getNumPartitions());
+
+      if (operator.getValueComparator() != null) {
+        // if we have a value comparator we need to secondary sort the values within key
+        reduced =
+            tuples
+                .mapToPair(t -> new Tuple2<>(new KeyedWindowValue<>(t._1, t._2), Empty.get()))
+                .repartitionAndSortWithinPartitions(
+                    partitioner, new SecondarySortComparator<>(operator.getValueComparator()))
+                .mapToPair(t -> new Tuple2<>(t._1.toKeyedWindow(), t._1.getValue()))
+                .mapPartitionsToPair(ReduceByKeyIterator::new)
+                .flatMapValues(new Reducer<>(reducer, accumulatorProvider));
+      } else {
+        // if the key is comparable, we can optimize for memory efficiency
+        reduced =
+            tuples
+                .repartitionAndSortWithinPartitions(partitioner)
+                .mapPartitionsToPair(ReduceByKeyIterator::new)
+                .flatMapValues(new Reducer<>(reducer, accumulatorProvider));
+      }
+    } else {
+      LOG.warn(
+          "Key for ["
+              + operator.getName()
+              + "] is not comparable, so we can not optimize for memory efficiency.");
+      // we can n
+      reduced = tuples.groupByKey().flatMapValues(new Reducer<>(reducer, accumulatorProvider));
     }
+
+    return reduced.map(
+        t -> {
+          final KeyedWindow<W, KEY> kw = t._1();
+          final OUT el = t._2();
+          return new SparkElement<>(kw.window(), kw.timestamp(), Pair.of(kw.key(), el));
+        });
```

Comment:
The `doTranslate()` method has grown to more than 100 lines and there are several nested structures. It hurts readability all together. New appropriately named methods may help.